### ISSUE TITLE
Envoy segmentation fault

### ISSUE THREAD

**Issue by iandyh** [istio/istio#25909]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**
Envoy had segmentation fault and exit with following exceptions:

![image](https://user-images.githubusercontent.com/431658/88632246-4bc1b000-d0ee-11ea-9401-86c47a9a0a63.png)

This is the log from one of the faulty istio-ingressgateway pods.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**
No segmentation fault

**Steps to reproduce the bug**
The segmentation fault happen from time to time, not very often. Traffic is light.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
control plane and data plane are both 1.6.4

**How was Istio installed?**
istioctl

**Environment where bug was observed (cloud vendor, OS, etc)**

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by PiotrSikora** [comment]
cc @lizan

**Comment by lizan** [comment]
@iandyh reached out via Envoy Slack, I need more details from you.

**Comment by iandyh** [comment]
@lizan Thanks. Will move the discussion to Slack.

**Comment by incfly** [comment]
For people who are potentially affected by this same issue, this should now should already be resolved in 1.7 release and future release. We might cherry-pick the fix to older release as well.

---

Debugging notes

## Reproduce issue

I setup the large number of namespaces for fortio load test, adapted from istio/tools repo. Istio is installed with default profile and options. I also need to use a script to constantly creating and deleting the namespace & services. Therefore the `ClusterEntry` destructor is then invoked, as showed in the stacktrace above.

This setup would then reproduce the issue after several hours running.

## Investigate

I build the istio/proxy with modified envoy and instrumentation. This narrows down the segfault in the callback manager's `remove`, https://github.com/istio/envoy/blob/fd7c70b3f3739554974769f1d0e72b706eb5c5a3/source/common/common/callback_impl.h#L59



```cpp
    bool found = std::find_if(callbacks_.begin(), callbacks_.end(),
                              [handle](const CallbackHolder& holder) -> bool {
                                return handle == &holder;
                              }) != callbacks_.end();
    ENVOY_LOG_MISC(error, "jianfeih callbacks_ remove found {}, size", found, callbacks_.size());
    callbacks_.remove_if(
        [handle](const CallbackHolder& holder) -> bool { return handle == &holder; });
```

It's interesting though since the `remove_if` predicate function only has read not de-reference of the iterator. Can't see how it get segfault.

In one of the ASAN build, I see a segfault with error information:

```
==15==ERROR: AddressSanitizer: heap-use-after-free on address 0x6060033b07e0 at pc 0x5615ffe5b711 bp 0x7fbbb09c6480 sp 0x7fbbb09c6478             
READ of size 8 at 0x6060033b07e0 thread T10                                                                                                           #0 0x5615ffe5b710  (/usr/local/bin/envoy+0x54993710)                                                                                          
    #1 0x5615ffe5a8ae  (/usr/local/bin/envoy+0x549928ae)

0x6060033b07e0 is located 0 bytes inside of 64-byte region [0x6060033b07e0,0x6060033b0820)                                                        freed by thread T9 here:                                                                                                                          
    #0 0x5615f4bbfccd  (/usr/local/bin/envoy+0x496f7ccd)                                                                                              #1 0x5615ffe55b44  (/usr/local/bin/envoy+0x5498db44)                                                                                          
    #2 0x5615ffe55af9  (/usr/local/bin/envoy+0x5498daf9)                                                                                              #3 0x5615ffe558ed  (/usr/local/bin/envoy+0x5498d8ed)
```
So basically one thread is doing remove_if, iterating the std::list
The other thread is might remove some iterator/element already.

The bottom of the callsite is HostImpl, owned by `ThreadLocalClusterManagerImpl`.

## Fix

Interesting in 1.7 and master of envoy, the callback manager implmenetatoin has been improved, so the ::remove() is doing removal precisely with a given iterator (stored in `CallbackHolder` when created).
https://github.com/envoyproxy/envoy/pull/11751/files

This way, remove() will no longer need to do iteration, removal on different thread won't cause race & conflict.

I validate this by running the master build of the Istio, with the same repro setup. The ingress gateway does not segfault ever.
